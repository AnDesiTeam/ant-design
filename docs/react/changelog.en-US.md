---
order: 10
title: Change Log
toc: false
---

<embed src="../../CHANGELOG.en-US.md"></embed>
