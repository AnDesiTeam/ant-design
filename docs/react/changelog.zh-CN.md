---
order: 10
title: 更新日志
toc: false
---

<embed src="../../CHANGELOG.zh-CN.md"></embed>
