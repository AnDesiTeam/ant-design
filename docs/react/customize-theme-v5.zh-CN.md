---
order: 7
title: 5.0 定制主题
---

Ant Design 设计规范和技术上支持灵活的样式定制，以满足业务和品牌上多样化的视觉需求，包括但不限于全局样式（主色、圆角、边框）和指定组件的视觉定制。

在 5.0 版本的 Ant Design 中，我们提供了一套全新的定制主题方案。不同于 4.x 版本的 less 和 CSS 变量，有了 CSS-in-JS 的加持后，动态主题的能力也得到了加强，包括但不限于：

1. 支持动态切换主题；
2. 支持同时存在多个主题；
3. 支持针对某个/某些组件主题变量；
4. ...

> TODO: 这里需要一张 GIF 展示动态主题能力

## Design Token

在正式介绍 5.0 定制主题的方案之前，我们需要对主题的组成部分有一个基本的了解。

与 4.x 版本相同的是，5.0 也提供了一批变量用于更改主题的各项参数。 在 4.x 中我们使用了 less 变量来实现定制主题的能力；而在 5.0 的 CSS-in-JS 方案中，我们使用 JS 变量作为动态主题的基础，这些变量我们统称为 **Design Token**。

### 基本组成

用过 4.x 定制主题的同学应该知道，在 4.x 中所有的 less 变量都是平铺的，`default.less` 这个文件就长达 1000+ 行。在 Design Token 中我们对原有的变量进行了精简，并且在设计师的引导下产出了一套更加贴合设计的三层结构，将 Design Token 拆解为 **Seed Token**、**Map Token** 和 **Alias Token** 三部分。

这三组 Token 并不是简单的分组，而是一个三层的派生关系，由 Seed Token 派生 Map Token，再由 Map Token 派生 Alias Token。这中间经历的过程实际上也是设计到代码的过程，Seed Token 和 Map Token 主要是由设计师产出并展开，命名逻辑会更偏向设计；到了 Alias Token 实际上已经是 Token 投入到代码中使用的环节，所以命名逻辑会逐渐偏向使用侧。

设计师对于这三种 Token 的定义是这样的：

- **Seed Token**：基础变量基础变量（Seed Token）意味着所有设计意图的起源。在 Ant Design 中，我们会基于 Seed Token 自动派生一套具有设计语义的梯度变量（Map Token）
- **Map Token**：梯度变量梯度变量（Map Token） 是基于 Seed 派生的梯度变量，我们精心设计的梯度变量模型具有良好的设计语义，可保证在亮暗色模式切换时保证视觉梯度的一致性。
- **Alias Token**：别名变量别名变量（Alias Token）是 Map Token 的别名。Alias Token 用于批量控制某些共性组件的样式。

### 基本算法

> TODO

### 演变过程

> TODO

### Component Token

除了整体的 Token 链路之外，各个组件也会开放自己的 Component Token 来实现针对组件的样式定制能力，不同的组件之间不会相互影响。一般来说会是一些具体场景下的 CSS 变量，我们会单独提出来作为 Component Token。

> TODO: 需要对各个组件补充 Component Token 的说明，这里可以链接到某个组件的页面

## 在代码中定制主题

### 使用 ConfigProvider

我们为 ConfigProvider 添加了 `theme` 属性，用于传递主题配置，你可以在这里任意修改在上文中提到的 Design Token 和 Component Token。

```tsx
import { ConfigProvider, Button } from 'antd';

export default () => (
  <ConfigProvider
    theme={{
      token: {
        colorPrimary: '#1890ff',
        radiusBase: 2,
      },
    }}
  >
    <Button />
  </ConfigProvider>
);
```

### API (暂定)

#### Theme

| 属性 | 类型 | 默认值 | 说明 |
| --- | --- | --- | --- |
| token | `SeedToken` | - | 用于修改 Seed Token |
| algorithm | `(token: SeedToken) => MapToken` \| `((token: SeedToken) => MapToken)[]` | `defaultAlgorithm` | 用于修改 Seed Token 到 Map Token 的算法 |
| override | OverrideToken | - | 用于修改 Map Token, Alias Token 和 Component Token |

#### OverrideToken

| 属性 | 类型 | 默认值 | 说明 |
| --- | --- | --- | --- |
| derivative | `MapToken` | - | 用于修改 Map Token |
| alias | `AliasToken` | - | 用于修改 Alias Token |
| `Component` (可以是任意 antd 组件名) | `ComponentToken & AliasToken` | - | 用于修改 Component Token 以及覆盖该组件消费的 Alias Token |

#### SeedToken (待补全)

| 属性         | 类型     | 默认值    | 说明     |
| ------------ | -------- | --------- | -------- |
| colorPrimary | `string` | `#1677ff` | 品牌主色 |
| radiusBase   | `number` | `6`       | 基础圆角 |

#### MapToken (待补全)

| 属性           | 类型     | 默认值    | 说明                                                 |
| -------------- | -------- | --------- | ---------------------------------------------------- |
| colorPrimaryBg | `string` | `#e6f4ff` | 主色的浅色背景颜色，一般用于视觉层级较弱的选中状态。 |

#### AliasToken (待补全)

| 属性                | 类型     | 默认值    | 说明                                     |
| ------------------- | -------- | --------- | ---------------------------------------- |
| controlItemBgActive | `string` | `#e6f4ff` | 用于控件类组件中的单项的激活态选中样式。 |

### 调试主题

我们提供了帮助用户调试主题的工具：[主题编辑器](https://ant-design.github.io/antd-token-previewer/~demos/docs-theme-editor-simple)

你可以使用此工具自由地修改 Design Token，以达到您对主题的期望。

## 主题展示

- [Ant Design 4.x 主题](https://ant-design.github.io/antd-token-previewer/~demos/docs-v4-theme)

> 可以引导用户贡献一些优质主题并收集起来

## FAQ

### 为什么 `theme` 从 `undefined` 变为对象或者变为 `undefined` 时组件重新 mount 了？

在 ConfigProvider 中我们通过 `DesignTokenContext` 传递 context，`theme` 为 `undefined` 时不会套一层 Provider，所以从无到有或者从有到无时 React 的 VirtualDOM 结构变化，导致组件重新 mount。解决方法：将 `undefined` 替换为空对象 `{}` 即可。
