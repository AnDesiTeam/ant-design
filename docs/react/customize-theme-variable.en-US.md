---
order: 7.1
title: Dynamic Theme (Experimental)
---

Besides [less customize theme](/docs/react/customize-theme), We also provide CSS Variable version to enable dynamic theme. You can check on [ConfigProvider](/components/config-provider/#components-config-provider-demo-theme) demo.

## Notice

This function need CSS Variable support which means it can not support IE. Please make sure your browser requirement.

## How to use

### Import antd.variable.min.css

Replace your import style file with CSS Variable version:

```diff
-- import 'antd/dist/antd.min.css';
++ import 'antd/dist/antd.variable.min.css';
```

Note: You need remove `babel-plugin-import` for the dynamic theme.

### Static config

Call ConfigProvider static function to modify theme color:

```ts
import { ConfigProvider } from 'antd';

ConfigProvider.config({
  theme: {
    primaryColor: '#25b864',
  },
});
```

## Conflict resolve

CSS Variable use `--ant` prefix by default. When multiple antd style file exist in your project, you can modify the prefix to fix it.

### Adjust

Modify `prefixCls` on the root of ConfigProvider:

```tsx
import { ConfigProvider } from 'antd';

export default () => (
  <ConfigProvider prefixCls="custom">
    <MyApp />
  </ConfigProvider>
);
```

You also need call the static function to modify `prefixCls`:

```ts
ConfigProvider.config({
  prefixCls: 'custom',
  theme: {
    primaryColor: '#25b864',
  },
});
```

### Compile less

Since prefix is modified. Origin `antd.variable.css` should also be replaced:

```bash
lessc --modify-var="ant-prefix=custom" antd/dist/antd.variable.less modified.css
```
