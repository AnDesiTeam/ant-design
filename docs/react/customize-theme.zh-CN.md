---
order: 5
title: 定制主题
---

Ant Design 设计规范上支持一定程度的样式定制，以满足业务和品牌上多样化的视觉需求，包括但不限于主色、圆角、边框和部分组件的视觉定制。

![](https://zos.alipayobjects.com/rmsportal/zTFoszBtDODhXfLAazfSpYbSLSEeytoG.png)

## 样式变量

antd 的样式使用了 [Less](http://lesscss.org/) 作为开发语言，并定义了一系列全局/组件的样式变量，你可以根据需求进行相应调整。

- [默认样式变量](https://github.com/ant-design/ant-design/blob/master/components/style/themes/default.less)

如果以上变量不能满足你的定制需求，可以给我们提 issue。

## 定制方式

我们使用 [modifyVars](http://lesscss.org/usage/#using-less-in-the-browser-modify-variables) 的方式来覆盖变量。
在具体工程实践中，有 `package.theme` 和 `less` 两种方案，选择一种即可。

可以在本地运行 [例子](https://github.com/ant-design/antd-init/tree/master/examples/customize-antd-theme) 查看定制效果。

### 1) theme 属性（推荐）

配置在 `package.json` 或 `.roadhogrc` 下的 `theme` 字段。theme 可以为配置为一个对象或文件路径。

```js
"theme": {
  "primary-color": "#1DA57A",
},
```

或者 [一个 js 文件](https://github.com/ant-design/antd-init/blob/master/examples/customize-antd-theme/theme.js)：

```js
"theme": "./theme.js",
```

定义 `theme` 属性时，需要配合使用（[antd-init](https://github.com/ant-design/antd-init) 或 [dva-cli](https://github.com/dvajs/dva-cli)。如果你使用的是其他脚手架，可以参考 [atool-build 中 less-loader 的 webpack 相关配置 ](https://github.com/ant-tool/atool-build/blob/a4b3e3eec4ffc09b0e2352d7f9d279c4c28fdb99/src/getWebpackCommonConfig.js#L131-L138)，利用 [less-loader](https://github.com/webpack/less-loader#less-options) 的 `modifyVars` 配置来覆盖原来的样式变量。

注意：

- 样式必须加载 less 格式。
  - 如果你在使用 [babel-plugin-import](https://github.com/ant-design/babel-plugin-import) 的 `style` 配置来引入样式，需要将配置值从 `'css'` 改为 `true`，这样会引入 less 文件。
  - 如果你是通过 `'antd/dist/antd.css'` 引入样式的，改为 `antd/dist/antd.less`。
- `dva-cli@0.7.0+` 的 `theme` 属性需要写在 [.roadhogrc](https://github.com/dvajs/dva-example-user-dashboard/commit/d6da33b3a6e18eb7f003752a4b00b5a660747c31) 文件里。
- 如果要覆盖 `@icon-url` 变量，内容需要包括引号 `"@icon-url": "'your-icon-font-path'"`（[修正示例](https://github.com/visvadw/dvajs-user-dashboard/pull/2)）。

#### 单独 webpack 中配置 theme

1.  [babel-plugin-import](https://github.com/ant-design/babel-plugin-import) 的 `style` 配置来引入样式，需要将配置值从 `'css'` 改为 `true`，这样会引入 less 文件

2. `package.json` 中配置 `theme` 字段，直接配置为对象：

   ```js
   "theme": {
     "primary-color": "#1DA57A",
   },
   ```

   或者配置为[一个 js 文件](https://github.com/ant-design/antd-init/blob/master/examples/customize-antd-theme/theme.js)：

   ```js
   "theme": "./theme.js",
   ```

   > 关于`primary-color`  还是 `@primary-color` 写法，两种都支持

3. `webpack.dev.config.js` 中读取 `theme` 的配置信息：

   ```
   const fs = require('fs')
   const pkgPath = path.resolve(__dirname, './package.json')
   const pkg = fs.existsSync(pkgPath) ? require(pkgPath) : {}
   let theme = {}
   if (pkg.theme && typeof pkg.theme === 'string') {
     let cfgPath = pkg.theme
     if (cfgPath.charAt(0) === '.') {
       cfgPath = path.resolve(__dirname, cfgPath)
     }
     const getThemeConfig = require(cfgPath)
     theme = getThemeConfig()
   } else if (pkg.theme && typeof pkg.theme === 'object') {
     theme = pkg.theme
   }
   ```

   根据配置文件的格式，上面的theme已经是个对象了，下面webpack，loader中可以直接使用

4.  `webpack.dev.config.json` 中关于less 处理的相关 loader 写法：

   ```
   module: {
     rules: [
       {  // 处理自己的less 文件，如果没用less组件可以删除掉
         test: /\.less$/, 
         exclude: path.resolve(__dirname, 'node_modules'),  
         use: ExtractTextPlugin.extract({
           fallback: 'style-loader',
           use: [
             { loader: 'css-loader' },
             {
               loader: 'less-loader',
             },
           ],
         }),
       },
       {
         test: /.less$/,
         include: path.resolve(__dirname, 'node_modules/antd'),  // 处理antd 组件的 less 必须有
         use: ExtractTextPlugin.extract({
           fallback: 'style-loader',
           use: [
             {
               loader: 'css-loader',
             },
             {
               loader: 'less-loader',
               options: {
                 sourceMap: true,
                 modules: false,
                 modifyVars: theme,
               },
             },
           ],
         }),
       },
     ],
   }
   ```

   以上使用的是 `webpack v 3.0` 的写法，可以根据不同版本调整 loader的写法

### 2) less

用 less 文件进行变量覆盖。

建立一个单独的 `less` 文件如下，再引入这个文件。

   ```css
   @import "~antd/dist/antd.less";   // 引入官方提供的 less 样式入口文件
   @import "your-theme-file.less";   // 用于覆盖上面定义的变量
   ```

注意：这种方式已经载入了所有组件的样式，不需要也无法和按需加载插件 `babel-plugin-import` 的 `style` 属性一起使用。

## 社区教程

- [How to Customize Ant Design with React & Webpack… the Missing Guide](https://medium.com/@GeoffMiller/how-to-customize-ant-design-with-react-webpack-the-missing-guide-c6430f2db10f)
