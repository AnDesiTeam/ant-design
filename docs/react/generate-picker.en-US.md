---
title: generate picker
skip: true
---

## How to use DatePicker with customize date library like dayjsï¼Ÿ

Consider of bundle size, you can replace momentjs with customize date library. We provide two ways to customize date library:

### Custom component

The first way is use `generatePicker` (or `generateCalendar`) helps to create Picker components.

For example:

#### DatePicker.tsx

```tsx
import { Dayjs } from 'dayjs';
import dayjsGenerateConfig from 'rc-picker/lib/generate/dayjs';
import generatePicker from 'antd/es/date-picker/generatePicker';
import 'antd/es/date-picker/style/index';

const DatePicker = generatePicker<Dayjs>(dayjsGenerateConfig);

export default DatePicker;
```

#### Calendar.tsx

```tsx
import { Dayjs } from 'dayjs';
import dayjsGenerateConfig from 'rc-picker/lib/generate/dayjs';
import generateCalendar from 'antd/es/calendar/generateCalendar';
import 'antd/es/calendar/style';

const Calendar = generateCalendar<Dayjs>(dayjsGenerateConfig);

export default Calendar;
```

Then use the custom component (`DatePicker, Calendar`) in page to replace antd's `DatePicker, Calendar`, use `<DatePicker picker="time" />` to replace antd's `<TimePicker/>`.

use

```js
import { DatePicker, Calendar } from 'src/components';
import format from 'dayjs';
```

instead of

```js
import { DatePicker, Calendar } from 'antd';
import format from 'moment';
```

If you are the user of [umi](https://umijs.org/), you can ref [antd4-use-dayjs-replace-moment](https://github.com/xiaohuoni/antd4-use-dayjs-replace-moment).

### Webpack plugin

We also provide another implementation. We provide `antd-dayjs-webpack-plugin` plugin to replace `momentjs` to `Day.js` directly without changing a line of existing code. More info at [antd-dayjs-webpack-plugin](https://github.com/ant-design/antd-dayjs-webpack-plugin).
