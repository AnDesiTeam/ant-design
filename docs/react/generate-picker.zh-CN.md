---
title: 自定义日期库
skip: true
---

## 如何在 DatePicker 中使用自定义日期库（如 dayjs ）？

考虑到包的大小，你可以用自定义日期库替换 momentjs。在这里我们提供了两种方式来实现替换:

### 自定义组件

第一种方法是使用`generatePicker`（或者：`generateCalendar` ）辅助创建 Picker 组件。

编写如下代码:

#### DatePicker.tsx

```tsx
import { Dayjs } from 'dayjs';
import dayjsGenerateConfig from 'rc-picker/lib/generate/dayjs';
import generatePicker from 'antd/es/date-picker/generatePicker';
import 'antd/es/date-picker/style/index';

const DatePicker = generatePicker<Dayjs>(dayjsGenerateConfig);

export default DatePicker;
```

#### Calendar.tsx

```tsx
import { Dayjs } from 'dayjs';
import dayjsGenerateConfig from 'rc-picker/lib/generate/dayjs';
import generateCalendar from 'antd/es/calendar/generateCalendar';
import 'antd/es/calendar/style';

const Calendar = generateCalendar<Dayjs>(dayjsGenerateConfig);

export default Calendar;
```

`TimePicker` 是带有 `picker` 属性的 `DatePicker` 的别名： `<TimePicker />` = `<DatePicker picker="timer" />`.

使用

```js
import { DatePicker, Calendar } from 'src/components';
import format from 'dayjs';
```

替换

```js
import { DatePicker, Calendar } from 'antd';
import format from 'moment';
```

如果你熟悉 [umi](https://umijs.org/)，那你可以参考 [antd4-use-dayjs-replace-moment](https://github.com/xiaohuoni/antd4-use-dayjs-replace-moment)。

### Webpack 配置替换

我们还提供另一种实现方式。使用 `antd-dayjs-webpack-plugin` 插件，无需对现有代码做任何修改直接替换成 `Day.js`。请参考 [antd-dayjs-webpack-plugin](https://github.com/ant-design/antd-dayjs-webpack-plugin)。
