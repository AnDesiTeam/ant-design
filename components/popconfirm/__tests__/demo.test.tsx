import * as React from 'react';
import Popconfirm from '..';
import demoTest, { rootPropsTest } from '../../../tests/shared/demoTest';

demoTest('popconfirm', {
  testRootProps: false,
});

rootPropsTest(
  'popconfirm',
  (props) => (
    <Popconfirm {...props}>
      <span />
    </Popconfirm>
  ),
  {
    findRootElements: () => document.querySelector('.ant-popover')!,
  },
);
