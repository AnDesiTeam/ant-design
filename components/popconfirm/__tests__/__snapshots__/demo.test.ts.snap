// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ./components/popconfirm/demo/async.tsx correctly 1`] = `
{
  "html": "<button type="button" class="ant-btn ant-btn-primary"><span>Open Popconfirm with async logic</span></button>",
  "type": "demo",
}
`;

exports[`renders ./components/popconfirm/demo/basic.tsx correctly 1`] = `
{
  "html": "<a href="#">Delete</a>",
  "type": "demo",
}
`;

exports[`renders ./components/popconfirm/demo/dynamic-trigger.tsx correctly 1`] = `
{
  "html": "<div><a href="#">Delete a task</a><br/><br/>Whether directly executeï¼š<button type="button" role="switch" aria-checked="true" class="ant-switch ant-switch-checked"><div class="ant-switch-handle"></div><span class="ant-switch-inner"><span class="ant-switch-inner-checked"></span><span class="ant-switch-inner-unchecked"></span></span></button></div>",
  "type": "demo",
}
`;

exports[`renders ./components/popconfirm/demo/icon.tsx correctly 1`] = `
{
  "html": "<a href="#">Delete</a>",
  "type": "demo",
}
`;

exports[`renders ./components/popconfirm/demo/locale.tsx correctly 1`] = `
{
  "html": "<a href="#">Delete</a>",
  "type": "demo",
}
`;

exports[`renders ./components/popconfirm/demo/placement.tsx correctly 1`] = `
{
  "html": "<div class="demo"><div style="margin-left:70px;white-space:nowrap"><button type="button" class="ant-btn ant-btn-default"><span>TL</span></button><button type="button" class="ant-btn ant-btn-default"><span>Top</span></button><button type="button" class="ant-btn ant-btn-default"><span>TR</span></button></div><div style="width:70px;float:left"><button type="button" class="ant-btn ant-btn-default"><span>LT</span></button><button type="button" class="ant-btn ant-btn-default"><span>Left</span></button><button type="button" class="ant-btn ant-btn-default"><span>LB</span></button></div><div style="width:70px;margin-left:304px"><button type="button" class="ant-btn ant-btn-default"><span>RT</span></button><button type="button" class="ant-btn ant-btn-default"><span>Right</span></button><button type="button" class="ant-btn ant-btn-default"><span>RB</span></button></div><div style="margin-left:70px;clear:both;white-space:nowrap"><button type="button" class="ant-btn ant-btn-default"><span>BL</span></button><button type="button" class="ant-btn ant-btn-default"><span>Bottom</span></button><button type="button" class="ant-btn ant-btn-default"><span>BR</span></button></div></div>",
  "type": "demo",
}
`;

exports[`renders ./components/popconfirm/demo/promise.tsx correctly 1`] = `
{
  "html": "<button type="button" class="ant-btn ant-btn-primary"><span>Open Popconfirm with Promise</span></button>",
  "type": "demo",
}
`;

exports[`renders ./components/popconfirm/demo/render-panel.tsx correctly 1`] = `
{
  "html": "<div class="ant-popover ant-popover-pure ant-popover-placement-top ant-popconfirm"><div class="ant-popover-content"><div class="ant-popover-arrow"></div><div class="ant-popover-inner" role="tooltip"><div class="ant-popover-inner-content"><div class="ant-popconfirm-inner-content"><div class="ant-popconfirm-message"><span class="ant-popconfirm-message-icon"><span role="img" aria-label="exclamation-circle" class="anticon anticon-exclamation-circle"><svg viewBox="64 64 896 896" focusable="false" data-icon="exclamation-circle" width="1em" height="1em" fill="currentColor" aria-hidden="true"><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"></path></svg></span></span><div class="ant-popconfirm-message-title ant-popconfirm-message-title-only">Are you OK?</div></div><div class="ant-popconfirm-description">Does this look good?</div><div class="ant-popconfirm-buttons"><button type="button" class="ant-btn ant-btn-default ant-btn-sm"><span>Cancel</span></button><button type="button" class="ant-btn ant-btn-primary ant-btn-sm"><span>OK</span></button></div></div></div></div></div></div><div class="ant-popover ant-popover-pure ant-popover-placement-bottomRight ant-popconfirm" style="width:250px"><div class="ant-popover-content" style="width:250px"><div class="ant-popover-arrow"></div><div class="ant-popover-inner" role="tooltip"><div class="ant-popover-inner-content"><div class="ant-popconfirm-inner-content"><div class="ant-popconfirm-message"><span class="ant-popconfirm-message-icon"><span role="img" aria-label="exclamation-circle" class="anticon anticon-exclamation-circle"><svg viewBox="64 64 896 896" focusable="false" data-icon="exclamation-circle" width="1em" height="1em" fill="currentColor" aria-hidden="true"><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"></path></svg></span></span><div class="ant-popconfirm-message-title ant-popconfirm-message-title-only">Are you OK?</div></div><div class="ant-popconfirm-description">Does this look good?</div><div class="ant-popconfirm-buttons"><button type="button" class="ant-btn ant-btn-default ant-btn-sm"><span>Cancel</span></button><button type="button" class="ant-btn ant-btn-primary ant-btn-sm"><span>OK</span></button></div></div></div></div></div></div>",
  "type": "demo",
}
`;

exports[`renders ./components/popconfirm/demo/wireframe.tsx correctly 1`] = `
{
  "html": "<div class="ant-popover ant-popover-pure ant-popover-placement-top ant-popconfirm"><div class="ant-popover-content"><div class="ant-popover-arrow"></div><div class="ant-popover-inner" role="tooltip"><div class="ant-popover-inner-content"><div class="ant-popconfirm-inner-content"><div class="ant-popconfirm-message"><span class="ant-popconfirm-message-icon"><span role="img" aria-label="exclamation-circle" class="anticon anticon-exclamation-circle"><svg viewBox="64 64 896 896" focusable="false" data-icon="exclamation-circle" width="1em" height="1em" fill="currentColor" aria-hidden="true"><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"></path></svg></span></span><div class="ant-popconfirm-message-title">Are you OK?</div></div><div class="ant-popconfirm-buttons"><button type="button" class="ant-btn ant-btn-default ant-btn-sm"><span>Cancel</span></button><button type="button" class="ant-btn ant-btn-primary ant-btn-sm"><span>OK</span></button></div></div></div></div></div></div><div class="ant-popover ant-popover-pure ant-popover-placement-bottomRight ant-popconfirm" style="width:250px"><div class="ant-popover-content" style="width:250px"><div class="ant-popover-arrow"></div><div class="ant-popover-inner" role="tooltip"><div class="ant-popover-inner-content"><div class="ant-popconfirm-inner-content"><div class="ant-popconfirm-message"><span class="ant-popconfirm-message-icon"><span role="img" aria-label="exclamation-circle" class="anticon anticon-exclamation-circle"><svg viewBox="64 64 896 896" focusable="false" data-icon="exclamation-circle" width="1em" height="1em" fill="currentColor" aria-hidden="true"><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm-32 232c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V296zm32 440a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"></path></svg></span></span><div class="ant-popconfirm-message-title">Are you OK?</div></div><div class="ant-popconfirm-buttons"><button type="button" class="ant-btn ant-btn-default ant-btn-sm"><span>Cancel</span></button><button type="button" class="ant-btn ant-btn-primary ant-btn-sm"><span>OK</span></button></div></div></div></div></div></div>",
  "type": "demo",
}
`;
